internal/services/users/user_data_source_test.go

```diff
@@ -32,6 +32,19 @@ func TestAccUserDataSource_byUserPrincipalNameNonexistent(t *testing.T) {
 	}})
 }
 
+// byUserPrincipalNameSpecialCharacters tries to find a user object by UPN.
+// This test case covers UPNs with special characters.
+// It is expected not to find the user.
+// It is not expected for it to fail with "Error: Finding user with UPN: \"[^\"]+\""
+func TestAccUserDataSource_byUserPrincipalNameSpecialCharacters(t *testing.T) {
+	data := acceptance.BuildTestData(t, "data.azuread_user", "test")
+
+	data.DataSourceTest(t, []resource.TestStep{{
+		Config:      UserDataSource{}.byUserPrincipalNameNonexistent(data),
+		ExpectError: regexp.MustCompile("User with UPN \"[^\"]+\" was not found"),
+	}})
+}
+
```

Comment:
This is great, but I think it would be nicer to look up an existing user as this will ensure our input is being processed as intended by the API.

We could update the `UserResource.complete()` test config to include an apostrophe in the UPN, and then probably omit this additional test for the data source.